text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+uninterrupted of each other, even on the monolith.  For instance in an ideal scenario changes to an integration
+platform code do not necessarily have to run the entire Sentry testsuite on every commit.  Likewise UI code
+ideally does not build the entirety of the Sentry UI codebase.
+
+### Import Compartmentalization
+
+As a first start imports in the Sentry codebase should be compartmentalized as much as possible.  Today we have
+some catch all modules like `sentry.models` which import everything.  This means that it's hard for as to track
+the actual dependencies from different pieces of code and also hides away circular imports.  One of the results of
+this is that if you run a test against a single pytest module, you still pull in the entire codebase which already
+takes a couple of seconds.  It also means that it's harder for test infrastructure code to analyze the minimal set
+of dependencies that might need testing.
+
+Likewise code in `sentry.utils` should most likely no longer import models etc.  Some of that code might event
+move out of `sentry` entirely.
+
+### Move Code from Getsentry to Senty
+
+Currently quite a bit of code lives in `getsentry` for reasons of assumed convenience.  However this has created
+the situation that it is quite easy to ship regressions because the changes in `getsentry` were not considered.
+There is likely a whole range of code that does not need to live in `getsentry` and moving it to `sentry` would
+make the testsuite more reliable, faster to run and catch more regressions early.
+
+### Remove Pickle in Database
+
+We still use pickled models in database code quite extensively.  This reliance on pickle makes it harder than
+necessary to read and write from this outside of the main Python codebase.  Moving this to JSON will simplify the
+handling of moving such code into environments where other code wants to access it.  We already require depickling
+in the data pipeline for such cases.
+
+### Phase out Buffers where Possible
+
+Buffers as they exist today could in many cases be replaced by better uses of Kafka and clickhouse.  Some buffers are
+not even as important any more for the product.  For instance the total counts for issues are already hidden from the
+more prominent places in the UI.  The issue with buffers is that they are rather hard to scale, require the use of
+pickle for the model updates and are hard to work with when it comes to filtering.  For instance we are able to give
+you the total event count on an issue, but not when broken down by some of the filters that the UI wants to provide
+(environment and release is possible, but any tag search will not able to give the right counts).
+
+## Relay
+
+Within relay some changes are likely to be necessary for continued to support of providing a stable service.  In
+particular as more and more work moves into Relay certain risks that the current architecture is carrying need to
```

Comment:
Is it a good idea that more and more work keeps moving into Relay ?
I see a lot of work that cannot be anywhere else like bucketing metrics and process dynamic sampling as we want customer relays to be able to drop transactions.
Though the more logic we put in Relay without a split between infra code and product code the harder it is to keep it stable.
Product logic that maybe should not be there can be:
- extract spans details (for the suspect spans feature)
- A various amount of replays specific processing logic 
- A lot of events specific normalization (the more we add envelope types the more product logic will end up in Relay to normalize the data structure)
- ...

Is there something that prevents us from making Relay simpler and moving all product logic into the specific data category ingestion pipeline to further isolate them?
Also did a bug specific to a data category processing ever cause an incident in relay that impacted others ?

Reply:
> Is it a good idea that more and more work keeps moving into Relay ?

Unclear, but I have been talking with @jan-auer already that if we keep doing we might want to split relay into two pieces two which are connected through a persistent queue. And once we do something like that we can also modularize it. Metrics aggregation I'm conflicted about, but in particular longer term there is a lot of reason to try to attempt symbolication before something hits the innermost parts of the pipeline, particularly now that we have a lot more mobile/native traffic.

> Is there something that prevents us from making Relay simpler and moving all product logic into the specific data category ingestion pipeline to further isolate them?

Anything we need to know ahead of dynamic sampling needs to be expanded/processes by relay.

Reply:
This needs a more concrete plan. We've been talking for a long time about doing symbolication in Relay but so far it's been a pie-in-the-sky idea.

I think the problems this pipeline-reordering brings up are mostly impossible to solve. I have a rough idea on how to do grouping-aware filtering in some cases, but I don't see how we can stackwalk before dynamic sampling without any access to our massive internal system symbols, for example.